tools/wake-format/formatter.h

```diff
@@ -0,0 +1,320 @@
+    while (!node.empty() && (node.id() == TOKEN_WS || node.id() == TOKEN_NL)) {
+      node.nextSiblingElement();
+    }
+  }
+};
+
+struct SpaceAction {
+  uint8_t count;
+  SpaceAction(uint8_t count) : count(count) {}
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    space(builder, count);
+  }
+};
+
+struct NewlineAction {
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    builder.append("\n");
+    space(builder, 4 * ctx.nest_level);
+  }
+};
+
+struct TokenAction {
+  uint8_t token_id;
+  TokenAction(uint8_t token_id) : token_id(token_id) {}
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    assert(node.id() == token_id);
+    builder.append(node.fragment().segment().str());
+    node.nextSiblingElement();
+  }
+};
+
+struct TokenReplaceAction {
+  uint8_t token_id;
+  const char* str;
+
+  TokenReplaceAction(uint8_t token_id, const char* str) : token_id(token_id), str(str) {}
+
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    assert(node.id() == token_id);
+    builder.append(str);
+    node.nextSiblingElement();
+  }
+};
+
+struct WhitespaceTokenAction : public TokenReplaceAction {
+  WhitespaceTokenAction() : TokenReplaceAction(TOKEN_WS, " ") {}
+};
+
+template <class Action1, class Action2>
+struct SeqAction {
+  Action1 action1;
+  Action2 action2;
+  SeqAction(Action1 a1, Action2 a2) : action1(a1), action2(a2) {}
+
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    action1.run(builder, ctx, node);
+    action2.run(builder, ctx, node);
+  }
+};
+
+template <class F>
+struct WalkAction {
+  F walker;
+  uint8_t node_type;
+
+  WalkAction(F walk, uint8_t node_type) : walker(walk), node_type(node_type) {}
+
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    assert(node.id() == node_type);
+    auto doc = walker(ctx.sub(builder), const_cast<const CSTElement&>(node));
+    builder.append(doc);
+    node.nextSiblingElement();
+  }
+};
+
+template <class F, class P>
+struct WalkPredicateAction {
+  F walker;
+  P predicate;
+
+  WalkPredicateAction(F walk, P p) : walker(walk), predicate(p) {}
+
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    assert(predicate(node.id()));
+    auto doc = walker(ctx.sub(builder), const_cast<const CSTElement&>(node));
+    builder.append(doc);
+    node.nextSiblingElement();
+  }
+};
+
+template <class FMT>
+struct IfAction {
+  FMT formatter;
+  uint8_t node_type;
+
+  IfAction(FMT formatter, uint8_t node_type) : formatter(formatter), node_type(node_type) {}
+
+  ALWAYS_INLINE void run(wcl::doc_builder& builder, ctx_t ctx, CSTElement& node) {
+    if (node.id() == node_type) {
+      builder.append(formatter.compose(ctx, node));
+    }
+  }
+};
+
+template <class IFMT, class EFMT>
+struct IfElseAction {
```

Comment:
This subsumes the above. You can just use an EpslionAction formatter when you only want the `if` in the Formatter API